File: java/pdfclown.lib/src/org/pdfclown/documents/contents/fonts/AfmParser.java
/*
  Copyright 2009-2015 Stefano Chizzolini. http://www.pdfclown.org

  Contributors:
    * Stefano Chizzolini (original code developer, http://www.stefanochizzolini.it)

  This file should be part of the source code distribution of "PDF Clown library"
  (the Program): see the accompanying README files for more info.

  This Program is free software; you can redistribute it and/or modify it under the terms
  of the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3 of the License, or (at your option) any later version.

  This Program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY,
  either expressed or implied; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.

  You should have received a copy of the GNU Lesser General Public License along with this
  Program (see README files); if not, go to the GNU website (http://www.gnu.org/licenses/).

  Redistribution and use, with or without modification, are permitted provided that such
  redistributions retain the above copyright notice, license and disclaimer, along with
  this list of conditions.
*/

package org.pdfclown.documents.contents.fonts;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.Hashtable;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
  AFM file format parser [AFM:4.1].

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @since 0.0.8
  @version 0.1.2.1, 05/05/15
*/
final class AfmParser
{
  // <class>
  // <classes>
  /**
    Font header (Global font information).
  */
  static final class FontMetrics
  {
    /**
      Whether the encoding is custom (symbolic font).
    */
    public int ascender;
    public int capHeight;
    public int descender;
    public String fontName;
    public boolean isCustomEncoding;
    public boolean isFixedPitch;
    public float italicAngle;
    public int stemH;
    public int stemV;
    public int underlinePosition;
    public int underlineThickness;
    public String weight;
    public int xHeight;
    public int xMax;
    public int xMin;
    public int yMax;
    public int yMin;
  }
  // </classes>

  // <dynamic>
  // <fields>
  public FontMetrics metrics;

  public Map<Integer,Integer> glyphIndexes;
  public Map<Integer,Integer> glyphKernings;
  public Map<Integer,Integer> glyphWidths;

  public BufferedReader fontData;//TODO: convert to IInputStream
  // </fields>

  // <constructors>
  AfmParser(
    BufferedReader fontData
    )
  {
    this.fontData = fontData;

    load();
  }
  // </constructors>

  // <interface>
  // <private>
  /**
    Loads the font data.
  */
  private void load(
    )
  {
    try
    {
      metrics = new FontMetrics();
      loadFontHeader();
      loadCharMetrics();
      loadKerningData();
    }
    catch(IOException e)
    {throw new RuntimeException(e);}
  }

  /**
    Loads the font header [AFM:4.1:3,4.1-4.4].

    @throws IOException
  */
  private void loadFontHeader(
    ) throws IOException
  {
    String line;
    Pattern linePattern = Pattern.compile("(\\S+)\\s+(.+)");
    while((line = fontData.readLine()) != null)
    {
      Matcher lineMatcher = linePattern.matcher(line);
      if(!lineMatcher.find())
        continue;

      String key = lineMatcher.group(1);
      if (key.equals("Ascender"))
      {metrics.ascender = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("CapHeight"))
      {metrics.capHeight = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("Descender"))
      {metrics.descender = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("EncodingScheme"))
      {metrics.isCustomEncoding = lineMatcher.group(2).equals("FontSpecific");}
      else if (key.equals("FontBBox"))
      {
        String[] coordinates = lineMatcher.group(2).split("\\s+");
        metrics.xMin = (int)Float.parseFloat(coordinates[0]);
        metrics.yMin = (int)Float.parseFloat(coordinates[1]);
        metrics.xMax = (int)Float.parseFloat(coordinates[2]);
        metrics.yMax = (int)Float.parseFloat(coordinates[3]);
      }
      else if(key.equals("FontName"))
      {metrics.fontName = lineMatcher.group(2);}
      else if (key.equals("IsFixedPitch"))
      {metrics.isFixedPitch = Boolean.parseBoolean(lineMatcher.group(2));}
      else if (key.equals("ItalicAngle"))
      {metrics.italicAngle = Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("StdHW"))
      {metrics.stemH = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("StdVW"))
      {metrics.stemV = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("UnderlinePosition"))
      {metrics.underlinePosition = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("UnderlineThickness"))
      {metrics.underlineThickness = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("Weight"))
      {metrics.weight = lineMatcher.group(2);}
      else if (key.equals("XHeight"))
      {metrics.xHeight = (int)Float.parseFloat(lineMatcher.group(2));}
      else if (key.equals("StartCharMetrics"))
      {break;}
    }
    if(metrics.ascender == 0)
    {metrics.ascender = metrics.yMax;}
    if(metrics.descender == 0)
    {metrics.descender = metrics.yMin;}
  }

  /**
    Loads individual character metrics [AFM:4.1:3,4,4.4,8].

    @throws IOException
  */
  private void loadCharMetrics(
    ) throws IOException
  {
    glyphIndexes = new Hashtable<Integer, Integer>();
    glyphWidths = new Hashtable<Integer,Integer>();

    String line;
    Pattern linePattern = Pattern.compile("C (\\S+) ; WX (\\S+) ; N (\\S+)");
    int implicitCharCode = Short.MAX_VALUE;
    while((line = fontData.readLine()) != null)
    {
      Matcher lineMatcher = linePattern.matcher(line);
      if(!lineMatcher.find())
      {
        if(line.equals("EndCharMetrics"))
          break;

        continue;
      }

      int charCode = Integer.parseInt(lineMatcher.group(1));
      int width = (int)Float.parseFloat(lineMatcher.group(2));
      String charName = lineMatcher.group(3);
      if(charCode < 0)
      {
        if(charName == null)
          continue;

        charCode = ++implicitCharCode;
      }
      int code = (
        charName == null
            || metrics.isCustomEncoding
          ? charCode
          : GlyphMapping.nameToCode(charName)
        );
      glyphIndexes.put(code,charCode);
      glyphWidths.put(charCode,width);
    }
  }

  /**
    Loads kerning data [AFM:4.1:3,4,4.5,9].

    @throws IOException
  */
  private void loadKerningData(
    ) throws IOException
  {
    glyphKernings = new Hashtable<Integer,Integer>();

    String line;
    while((line = fontData.readLine()) != null)
    {
      if(line.startsWith("StartKernPairs"))
        break;
    }

    Pattern linePattern = Pattern.compile("KPX (\\S+) (\\S+) (\\S+)");
    while((line = fontData.readLine()) != null)
    {
      Matcher lineMatcher = linePattern.matcher(line);
      if(!lineMatcher.find())
      {
        if(line.equals("EndKernPairs"))
          break;

        continue;
      }

      int code1 = GlyphMapping.nameToCode(lineMatcher.group(1));
      int code2 = GlyphMapping.nameToCode(lineMatcher.group(2));
      int pair = code1 << 16 + code2;
      int value = (int)Float.parseFloat(lineMatcher.group(3));

      glyphKernings.put(pair,value);
    }
  }
  // </private>
  // </interface>
  // </dynamic>
  // </class>
}

File: PdfClown-NetStandard/PdfClown/org/pdfclown/documents/contents/fonts/AfmParser.cs
/*
  Copyright 2009-2015 Stefano Chizzolini. http://www.pdfclown.org

  Contributors:
    * Stefano Chizzolini (original code developer, http://www.stefanochizzolini.it)

  This file should be part of the source code distribution of "PDF Clown library" (the
  Program): see the accompanying README files for more info.

  This Program is free software; you can redistribute it and/or modify it under the terms
  of the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3 of the License, or (at your option) any later version.

  This Program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY,
  either expressed or implied; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.

  You should have received a copy of the GNU Lesser General Public License along with this
  Program (see README files); if not, go to the GNU website (http://www.gnu.org/licenses/).

  Redistribution and use, with or without modification, are permitted provided that such
  redistributions retain the above copyright notice, license and disclaimer, along with
  this list of conditions.
*/

using org.pdfclown.bytes;
using org.pdfclown.util;

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace org.pdfclown.documents.contents.fonts
{
  /**
    <summary>AFM file format parser [AFM:4.1].</summary>
  */
  public sealed class AfmParser
  {
    #region types
    /**
      <summary>Font header (Global font information).</summary>
    */
    public sealed class FontMetrics
    {
      public int Ascender;
      public int CapHeight;
      public int Descender;
      public string FontName;
      public bool IsCustomEncoding;
      public bool IsFixedPitch;
      public float ItalicAngle;
      public int UnderlinePosition;
      public int UnderlineThickness;
      public int StemH;
      public int StemV;
      public string Weight;
      public int XHeight;
      public int XMax;
      public int XMin;
      public int YMax;
      public int YMin;
    }
    #endregion

    #region dynamic
    #region fields
    public FontMetrics Metrics;

    public Dictionary<int,int> GlyphIndexes;
    public Dictionary<int,int> GlyphKernings;
    public Dictionary<int,int> GlyphWidths;

    public IInputStream FontData;
    #endregion

    #region constructors
    internal AfmParser(
      IInputStream fontData
      )
    {
      FontData = fontData;

      Load();
    }
    #endregion

    #region interface
    #region private
    private void Load(
      )
    {
      Metrics = new FontMetrics();
      LoadFontHeader();
      LoadCharMetrics();
      LoadKerningData();
    }

    /**
      <summary>Loads the font header [AFM:4.1:3,4.1-4.4].</summary>
    */
    private void LoadFontHeader(
      )
    {
      string line;
      Regex linePattern = new Regex("(\\S+)\\s+(.+)");
      while((line = FontData.ReadLine()) != null)
      {
        MatchCollection lineMatches = linePattern.Matches(line);
        if(lineMatches.Count < 1)
          continue;

        Match lineMatch = lineMatches[0];

        string key = lineMatch.Groups[1].Value;
        if(key.Equals("Ascender"))
        {Metrics.Ascender = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("CapHeight"))
        {Metrics.CapHeight = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("Descender"))
        {Metrics.Descender = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("EncodingScheme"))
        {Metrics.IsCustomEncoding = lineMatch.Groups[2].Value.Equals("FontSpecific");}
        else if(key.Equals("FontBBox"))
        {
          string[] coordinates = Regex.Split(lineMatch.Groups[2].Value,"\\s+");
          Metrics.XMin = ConvertUtils.ParseAsIntInvariant(coordinates[0]);
          Metrics.YMin = ConvertUtils.ParseAsIntInvariant(coordinates[1]);
          Metrics.XMax = ConvertUtils.ParseAsIntInvariant(coordinates[2]);
          Metrics.YMax = ConvertUtils.ParseAsIntInvariant(coordinates[3]);
        }
        else if(key.Equals("FontName"))
        {Metrics.FontName = lineMatch.Groups[2].Value;}
        else if(key.Equals("IsFixedPitch"))
        {Metrics.IsFixedPitch = Boolean.Parse(lineMatch.Groups[2].Value);}
        else if(key.Equals("ItalicAngle"))
        {Metrics.ItalicAngle = ConvertUtils.ParseFloatInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("StdHW"))
        {Metrics.StemH = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("StdVW"))
        {Metrics.StemV = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("UnderlinePosition"))
        {Metrics.UnderlinePosition = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("UnderlineThickness"))
        {Metrics.UnderlineThickness = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("Weight"))
        {Metrics.Weight = lineMatch.Groups[2].Value;}
        else if(key.Equals("XHeight"))
        {Metrics.XHeight = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);}
        else if(key.Equals("StartCharMetrics"))
          break;
      }
      if(Metrics.Ascender == 0)
      {Metrics.Ascender = Metrics.YMax;}
      if(Metrics.Descender == 0)
      {Metrics.Descender = Metrics.YMin;}
    }

    /**
      <summary>Loads individual character metrics [AFM:4.1:3,4,4.4,8].</summary>
    */
    private void LoadCharMetrics(
      )
    {
      GlyphIndexes = new Dictionary<int,int>();
      GlyphWidths = new Dictionary<int,int>();

      string line;
      Regex linePattern = new Regex("C (\\S+) ; WX (\\S+) ; N (\\S+)");
      int implicitCharCode = short.MaxValue;
      while((line = FontData.ReadLine()) != null)
      {
        MatchCollection lineMatches = linePattern.Matches(line);
        if(lineMatches.Count < 1)
        {
          if(line.Equals("EndCharMetrics"))
            break;

          continue;
        }

        Match lineMatch = lineMatches[0];

        int charCode = ConvertUtils.ParseIntInvariant(lineMatch.Groups[1].Value);
        int width = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[2].Value);
        string charName = lineMatch.Groups[3].Value;
        if(charCode < 0)
        {
          if(charName == null)
            continue;

          charCode = ++implicitCharCode;
        }
        int code = (
          charName == null
              || Metrics.IsCustomEncoding
            ? charCode
            : GlyphMapping.NameToCode(charName).Value
          );

        GlyphIndexes[code] = charCode;
        GlyphWidths[charCode] = width;
      }
    }

    /**
      <summary>Loads kerning data [AFM:4.1:3,4,4.5,9].</summary>
    */
    private void LoadKerningData(
      )
    {
      GlyphKernings = new Dictionary<int,int>();

      string line;
      while((line = FontData.ReadLine()) != null)
      {
        if(line.StartsWith("StartKernPairs"))
          break;
      }

      Regex linePattern = new Regex("KPX (\\S+) (\\S+) (\\S+)");
      while((line = FontData.ReadLine()) != null)
      {
        MatchCollection lineMatches = linePattern.Matches(line);
        if(lineMatches.Count < 1)
        {
          if(line.Equals("EndKernPairs"))
            break;

          continue;
        }

        Match lineMatch = lineMatches[0];

        int code1 = GlyphMapping.NameToCode(lineMatch.Groups[1].Value).Value;
        int code2 = GlyphMapping.NameToCode(lineMatch.Groups[2].Value).Value;
        int pair = code1 << 16 + code2;
        int value = ConvertUtils.ParseAsIntInvariant(lineMatch.Groups[3].Value);

        GlyphKernings[pair] = value;
      }
    }
    #endregion
    #endregion
    #endregion
  }
}